File: 108.convert-sorted-array-to-binary-search-tree.go
/*
 * @lc app=leetcode id=108 lang=golang
 *
 * [108] Convert Sorted Array to Binary Search Tree
 */

//  nums = [-10, -3, 0, 5, 9]
//  輸出:
//  [0 -3 9 -10 nil 5]

// 高度平衡二叉搜索樹 (BST):

// 二叉搜索樹的性質：左子樹的所有節點值小於根節點，右子樹的所有節點值大於根節點。
// 高度平衡：二叉樹中每個節點的左右子樹高度差不超過 1。

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func sortedArrayToBST(nums []int) *TreeNode {
	// 定義遞歸函數
	var helper func(left, right int) *TreeNode
	helper = func(left, right int) *TreeNode {
		if left > right {
			return nil
		}

		// 選擇中間元素作為根節點
		mid := left + (right-left)/2
		root := &TreeNode{Val: nums[mid]}

		// 遞歸構建左子樹和右子樹
		root.Left = helper(left, mid-1)
		root.Right = helper(mid+1, right)

		return root
	}

	// 初始調用遞歸函數
	return helper(0, len(nums)-1)
}

// @lc code=end

File: 108.convert-sorted-array-to-binary-search-tree.py
#
# @lc app=leetcode id=108 lang=python
#
# [108] Convert Sorted Array to Binary Search Tree
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def sortedArrayToBST(self, nums):
        """
        :type nums: List[int]
        :rtype: Optional[TreeNode]
        """
        #BST is a binary search tree 
        # It is a binary tree where each node has a value greater than the values in its left subtree and less than the values in its right subtree
        if not nums:
            return None

        #Find the middle index
        mid = len(nums) // 2

        #Create the root node
        root = TreeNode(nums[mid])

        #Recursively build the left and right subtrees
        #Left subtree process the first half of the array, with indices from 0 to mid-1
        root.left = self.sortedArrayToBST(nums[:mid])
        #Right subtree process the left half 0f the array, with indices from mid + 1 to the end
        root.right = self.sortedArrayToBST(nums[mid + 1:])

        return root

# @lc code=end

